File: nhitomi-idx/nhitomi/Discord/Commands/BookListMessage.cs
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using nhitomi.Database;
using nhitomi.Localization;
using nhitomi.Scrapers;

namespace nhitomi.Discord.Commands
{
    public class BookListMessage : InteractiveMessage, IListTriggerTarget
    {
        public INavigableAsyncEnumerator<DbBook> Enumerator { get; set; }
        public (DbBook book, DbBookContent content)? Current { get; private set; }

        readonly nhitomiCommandContext _context;
        readonly ILocale _l;
        readonly ILinkGenerator _link;
        readonly IBookContentSelector _selector;
        readonly IScraperService _scrapers;

        public BookListMessage(nhitomiCommandContext context, ILinkGenerator link, IBookContentSelector selector, IScraperService scrapers)
        {
            _context  = context;
            _l        = context.Locale.Sections["get.book"];
            _link     = link;
            _selector = selector;
            _scrapers = scrapers;
        }

        protected override IEnumerable<ReactionTrigger> CreateTriggers()
        {
            foreach (var trigger in base.CreateTriggers())
                yield return trigger;

            yield return new BookReadTrigger(_context, () => Current);
            yield return new ListTrigger(this, ListTriggerDirection.Left);
            yield return new ListTrigger(this, ListTriggerDirection.Right);
            yield return new DestroyTrigger(this);
        }

        bool _valid;

        protected override async Task OnInitialize(CancellationToken cancellationToken = default)
        {
            await base.OnInitialize(cancellationToken);

            _valid = await Enumerator.MoveNextAsync();
        }

        protected override async Task<ReplyContent> RenderAsync(CancellationToken cancellationToken = default)
        {
            // todo: no results message
            if (!_valid)
                return null;

            var (book, content) = Current ??= (Enumerator.Current, await _selector.SelectAsync(Enumerator.Current, _context, cancellationToken));

            return BookMessage.Render(book, content, _l, _link, _scrapers);
        }

        int IListTriggerTarget.Position => Enumerator.Position;

        public async Task<bool> SetPositionAsync(int position, CancellationToken cancellationToken = default)
        {
            if (await Enumerator.MoveToAsync(position))
            {
                Current = null;
                return true;
            }

            return false;
        }

        protected override async ValueTask OnDisposeAsync()
        {
            await base.OnDisposeAsync();
            await Enumerator.DisposeAsync();
        }
    }
}

File: nhitomi-discord/src/Messages/BookListMessage.ts
import { AsyncArray } from "../asyncArray";
import { Book, ObjectType, SpecialCollection } from "nhitomi-api";
import { BookMessage } from "./BookMessage";
import { ReadTrigger } from "../Triggers/ReadTrigger";
import { SourcesTrigger } from "../Triggers/SourcesTrigger";
import { ListTrigger } from "../Triggers/ListTrigger";
import { DestroyTrigger } from "../Triggers/DestroyTrigger";
import { InteractiveMessage, RenderResult } from "../Interactive/message";
import { ReactionTrigger } from "../Interactive/trigger";
import { FavoriteTrigger } from "../Triggers/FavoriteTrigger";
import { MessageContext } from "../context";

export class BookListMessage extends InteractiveMessage {
  position = 0;
  book: Book;

  get content() {
    return (
      this.book.contents.find((content) => content.language === this.context.user.language) || this.book.contents[0]
    );
  }

  get favoriteObject() {
    return {
      id: this.book.id,
      name: this.book.primaryName,
    };
  }

  constructor(context: MessageContext, readonly items: AsyncArray<Book>, initial: Book) {
    super(context);

    this.book = initial;
  }

  protected async render() {
    const current = await this.items.get(this.position);

    if (current) {
      this.book = current;
    } else {
      this.position = Math.max(0, Math.min(this.items.cachedLength - 1, this.position));
    }

    return this.processRenderResult(BookMessage.renderStatic(this.context.locale, this.book, this.content));
  }

  protected processRenderResult(result: RenderResult): RenderResult {
    return result;
  }

  protected createTriggers(): ReactionTrigger[] {
    if (!this.book) return [];

    return [
      ...super.createTriggers(),

      new FavoriteTrigger(this, ObjectType.Book, SpecialCollection.Favorites),
      new ReadTrigger(this),
      new SourcesTrigger(this),
      new ListTrigger(this, "left"),
      new ListTrigger(this, "right"),
      new DestroyTrigger(this),
    ];
  }
}
